### PR TITLE
feat: code cleanup and improving existing w3c validations

### PR DESCRIPTION
**Pull request by gjj**
Moved the following to `shared`:

- `@types`
- `merkle`
- `serialize`
- `utils`
- `validate`, though it contains `v3`-specific validations (RFC3339 datetime and RFC3986 URI)
- `logger.ts`

Moved the following to `v2`:

- `digest`, as the `digestDocument` between `v2` and `v3` is a little bit different
- `salt`
- `schema`

Todo:

* [ ] Fix `utils` and `e2e.v3` tests
* [ ] Look into `compact` and `jsonld`
* [ ] Validating a `VerifiableCredential`'s `type` values using `expansionMap`, see https://github.com/w3c/vc-test-suite/issues/97 and https://github.com/digitalbazaar/jsonld.js/issues/398
* [ ] Move `sign` to `shared` once it has been implemented for `v3`, see #95 
* [x] Fix CI path since paths have now changed

OA v3 Schema

| Property            | Standard            | Description                                                                                                                                                                                                                                                                                                                                  | Required |
|---------------------|---------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|
| `@context`          | W3C VC DM 1.0       | Can be `string` or `array`, but if it is an `array`, the first item MUST be a URI of value `https://www.w3.org/2018/credentials/v1`.                                                                                                                                                                                                         | ✅        |
| `issuanceDate`      | W3C VC DM 1.0       | Must be a string value of an RFC3339 combined date and time string representing the date and time the credential becomes valid, which could be a date and time in the future.                                                                                                                                                                | ✅        |
| `issuer`            | W3C VC DM 1.0       | Can be a `string` or an `object`. If it a `string`, it must be a URI. But if it is an `object`, it must contain an `id`  property. It is recommended that the URI, when dereferenced, results in a document containing machine-readable information about the issuer that can be used to verify the information expressed in the credential. | ✅        |
| `type`              | W3C VC DM 1.0       | Can be a `string` or an `array`. If it is a `string`, it must be `VerifiableCredential`. If it is an `array`, the first item should be `VerifiableCredential`, and the remaining items must map to one or more URIs defined in `@context`.                                                                                                   | ✅        |
| `credentialSubject` | W3C VC DM 1.0       | Can be an `object` or an `array` of `object`s, which can contain multiple subjects.                                                                                                                                                                                                                                                          | ✅        |
| `proof`             | W3C VC DM 1.0       |                                                                                                                                                                                                                                                                                                                                              | ✅        |
| `name`              | OpenAttestation 3.0 |                                                                                                                                                                                                                                                                                                                                              | No       |
| `reference`         | OpenAttestation 3.0 |                                                                                                                                                                                                                                                                                                                                              | No       |
| `recipient`         | OpenAttestation 3.0 |                                                                                                                                                                                                                                                                                                                                              | No       |
| `template`          | OpenAttestation 3.0 |                                                                                                                                                                                                                                                                                                                                              | ✅        |
| `validFrom`         | W3C VC DM 1.1       | Reserved for a future version of VC Data Model.                                                                                                                                                                                                                                                                                              | No       |
| `issued`            | W3C VC DM 1.1       | Reserved for a future version of VC Data Model.                                                                                                                                                                                                                                                                                              | No       |

### CODE REVIEW

**Comment by AdamJLemmon** [comment]
@gjj noticed the note of moving `sign` to `v2` as it is only used there. Although that is currently the case I just want to make sure that it is still supported as part of `v3` as it comes together.   So v3 I think should include `sign`, prior to release if not at present.  Just a comment I wanted to highlight but will defer to you guys as to how you'd like to manage migration to v3 :)  Great stuff.

**Comment by gjj** [comment]
@AdamJLemmon thanks for the highlight! Will move it back to `shared`.

**Comment by Nebulis** [comment]
@AdamJLemmon we will handle that in subsequent MR, could you open an issue and tag it with w3c label ?

File: src/shared/sign/index.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export * from "./sign";
```
Comment (Nebulis) [inline src/shared/sign/index.ts:1]: to move to v2 as discussed

File: src/shared/validate/validate.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+const logger = getLogger("validate");
+
+export const validateSchema = (document: any, validator: Ajv.ValidateFunction): Ajv.ErrorObject[] => {
+  if (!validator) {
+    throw new Error("No schema validator provided");
+  }
+  const valid = validator(document.version === SchemaId.v3 ? document : getData(document));
+  // console.log(validator.errors);
+  if (!valid) {
+    console.log({ validateError: validator.errors });
```
Comment (Nebulis) [inline src/shared/validate/validate.ts:20]: to remove

File: src/shared/validate/validate.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      console.log(info);
+      if (info.unmappedProperty) {
+        throw new Error("The property '" + info.unmappedProperty + "' in the input was not defined in the context");
+      }
+    }
+  });
+
+  // console.log(compacted);
+
+  if (validateTypeWithContext) {
```
Comment (Nebulis) [inline src/shared/validate/validate.ts:131]: to remove

File: src/v2/digest/digest.test.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
-import { digestDocument, flattenHashArray } from "./digest";
-import { SchematisedDocument, SchemaId } from "../@types/document";
+import { digestDocument as digestDocumentV2, flattenHashArray } from "./digest";
```
Comment (Nebulis) [inline src/v2/digest/digest.test.ts:1]: dont need to rename here

File: src/v2/verify.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { get } from "lodash";
+import { keccak256 } from "js-sha3";
+import { digestDocument as digestDocumentV2 } from "./digest";
```
Comment (Nebulis) [inline src/v2/verify.ts:3]: dont need to rename here

File: src/v2/wrap.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import { digestDocument as digestDocumentV2 } from "./digest";
```
Comment (Nebulis) [inline src/v2/wrap.ts:1]: dont need to rename

File: src/v3/digest/digest.test.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import { digestDocument as digestDocumentV3 } from "./digest";
```
Comment (Nebulis) [inline src/v3/digest/digest.test.ts:1]: dont need to rename

File: src/shared/sign/index.ts (hunk 1 above)
Comment (gjj) [reply src/shared/sign/index.ts]: Moved to `v2` in 608a779, will shift it back to `shared` once it has been implemented for `v3`.

File: src/shared/validate/validate.ts (hunk 2 above)
Comment (gjj) [reply src/shared/validate/validate.ts]: removed!

File: src/v2/digest/digest.test.ts (hunk 4 above)
Comment (gjj) [reply src/v2/digest/digest.test.ts]: removed the unnecessary rename

File: src/v3/digest/digest.test.ts (hunk 7 above)
Comment (gjj) [reply src/v3/digest/digest.test.ts]: removed in bba31a6

File: src/v2/wrap.ts (hunk 6 above)
Comment (gjj) [reply src/v2/wrap.ts]: removed in bba31a6

File: src/v2/verify.ts (hunk 5 above)
Comment (gjj) [reply src/v2/verify.ts:3]: removed in bba31a6

Comment (gjj) [reply src/v2/verify.ts]: removed unnecessary renames in c6a0ce8

File: src/shared/validate/validate.ts (hunk 3 above)
Comment (gjj) [reply src/shared/validate/validate.ts]: removed the `validateTypeWithContext` option in d67df54

**Pull request merged**